tests/datetime_test.py

```diff
@@ -87,9 +88,82 @@ def test_date_can_handle_datetime(self):
         ]
         self.session.query(self.Character).first()
 
-    def test_date_cannot_handle_tz_aware_datetime(self):
+    def test_date_can_handle_tz_aware_datetime(self):
         character = self.Character()
         character.name = "Athur"
         character.timestamp = datetime(2009, 5, 13, 19, 19, 30, tzinfo=CST())
         self.session.add(character)
-        self.assertRaises(DBAPIError, self.session.commit)
+
+
+Base = declarative_base()
+
+
+class FooBar(Base):
+    __tablename__ = "foobar"
+    name = sa.Column(sa.String, primary_key=True)
+    date = sa.Column(sa.Date)
+    datetime = sa.Column(sa.DateTime)
+
+
+@pytest.fixture
+def session(cratedb_service):
+    engine = cratedb_service.database.engine
+    session = sessionmaker(bind=engine)()
+
+    Base.metadata.drop_all(engine, checkfirst=True)
+    Base.metadata.create_all(engine, checkfirst=True)
+    return session
+
+
+@pytest.mark.skip("Test case not supported on SQLAlchemy 1.3")
+def test_datetime_notz(session):
+    """
+    An integration test for `sa.Date` and `sa.DateTime`, not using timezones.
+    """
+
+    # Insert record.
+    foo_item = FooBar(
+        name="foo",
+        date=dt.date(2009, 5, 13),
+        datetime=dt.datetime(2009, 5, 13, 19, 19, 30),
+    )
+    session.add(foo_item)
+    session.commit()
+    session.execute(sa.text("REFRESH TABLE foobar"))
+
+    # Query record.
+    result = session.execute(sa.select(FooBar.name, FooBar.date, FooBar.datetime)).mappings().first()
+
+    # Compare outcome.
+    assert result["date"].year == 2009
+    assert result["datetime"].year == 2009
+    assert result["datetime"].tzname() is None
+    assert result["datetime"].timetz() == dt.time(19, 19, 30)
+    assert result["datetime"].tzinfo is None
+
+
+@pytest.mark.skip("Test case not supported on SQLAlchemy 1.3")
+def test_datetime_tz(session):
+    """
+    An integration test for `sa.Date` and `sa.DateTime`, now using timezones.
+    """
+
+    # Insert record.
+    foo_item = FooBar(
+        name="foo",
+        date=dt.date(2009, 5, 13),
+        datetime=dt.datetime(2009, 5, 13, 19, 19, 30, tzinfo=CST()),
+    )
+    session.add(foo_item)
+    session.commit()
+    session.execute(sa.text("REFRESH TABLE foobar"))
+
+    # Query record.
+    result = session.execute(sa.select(FooBar.name, FooBar.date, FooBar.datetime)).mappings().first()
+
+    # Compare outcome.
+    assert result["date"].year == 2009
+    assert result["datetime"].year == 2009
+    assert result["datetime"].tzname() is None
+    assert result["datetime"].timetz() == dt.time(19, 19, 30)
```

Comment:
> When inserting or updating records using timezone-aware Python `date` or `datetime` objects, timezone information will not be preserved.

-- https://cratedb.com/docs/sqlalchemy-cratedb/data-types.html

> Shouldn't it be converted to UTC first?

I am not sure. We may want to validate how it works with PostgreSQL and its SQLAlchemy dialect?

Reply:
We need to use `timestamp WITH time zone` as the column data type to make the conversion, otherwise the time zone information is just stripped:
```
cr> create table test(id string, tz timestamp with time zone);
CREATE OK, 1 row affected (0.359 sec)
cr> insert into test values('UTC1', '2024-06-24 10:20:30.123UTC');
INSERT OK, 1 row affected (0.017 sec)
cr> insert into test values('UTC2', '2024-06-24 10:20:30.123');
INSERT OK, 1 row affected (0.017 sec)
cr> insert into test values('+2h', '2024-06-24 10:20:30.123+02:00');
INSERT OK, 1 row affected (0.009 sec)
cr> insert into test values('+3h', '2024-06-24 10:20:30.123+03:00');
INSERT OK, 1 row affected (0.018 sec)
cr> select * from test order by id;
+------+---------------+
| id   |            tz |
+------+---------------+
| +2h  | 1719217230123 |
| +3h  | 1719213630123 |
| UTC1 | 1719224430123 |
| UTC2 | 1719224430123 |
+------+---------------+
SELECT 4 rows in set (0.004 sec)
cr> drop table test;
DROP OK, 1 row affected (0.045 sec)
cr> create table test(id string, tz timestamp without time zone);
CREATE OK, 1 row affected (0.361 sec)
cr> insert into test values('UTC1', '2024-06-24 10:20:30.123UTC');
INSERT OK, 1 row affected (0.017 sec)
cr> insert into test values('UTC2', '2024-06-24 10:20:30.123');
INSERT OK, 1 row affected (0.017 sec)
cr> insert into test values('+2h', '2024-06-24 10:20:30.123+02:00');
INSERT OK, 1 row affected (0.018 sec)
cr> insert into test values('+3h', '2024-06-24 10:20:30.123+03:00');
INSERT OK, 1 row affected (0.017 sec)
cr> select * from test order by id;
+------+---------------+
| id   |            tz |
+------+---------------+
| +2h  | 1719224430123 |
| +3h  | 1719224430123 |
| UTC1 | 1719224430123 |
| UTC2 | 1719224430123 |
+------+---------------+
SELECT 4 rows in set (0.004 sec)
```

Can we have a test `with time zone` for the col type?

Reply:
fyi, we follow the behavior of postgres:
```
matriv=> create table test(id text, tz timestamp without time zone);
CREATE TABLE
matriv=> insert into test values('UTC1', '2024-06-24 10:20:30.123UTC');
INSERT 0 1
matriv=> insert into test values('UTC2', '2024-06-24 10:20:30.123');
INSERT 0 1
matriv=> insert into test values('+2hr', '2024-06-24 10:20:30.123+02:00');
INSERT 0 1
matriv=> insert into test values('+3hr', '2024-06-24 10:20:30.123+03:00');
INSERT 0 1
matriv=> select * from test;
  id  |           tz            
------+-------------------------
 UTC1 | 2024-06-24 10:20:30.123
 UTC2 | 2024-06-24 10:20:30.123
 +2hr | 2024-06-24 10:20:30.123
 +3hr | 2024-06-24 10:20:30.123
(4 rows)

matriv=> drop table test;
DROP TABLE
matriv=> create table test(id text, tz timestamp with time zone);
CREATE TABLE
matriv=> insert into test values('UTC1', '2024-06-24 10:20:30.123UTC');
INSERT 0 1
matriv=> insert into test values('UTC2', '2024-06-24 10:20:30.123');
INSERT 0 1
matriv=> insert into test values('+2hr', '2024-06-24 10:20:30.123+02:00');
INSERT 0 1
matriv=> insert into test values('+3hr', '2024-06-24 10:20:30.123+03:00');
INSERT 0 1
matriv=> select * from test;
  id  |             tz             
------+----------------------------
 UTC1 | 2024-06-24 13:20:30.123+03
 UTC2 | 2024-06-24 10:20:30.123+03
 +2hr | 2024-06-24 11:20:30.123+03
 +3hr | 2024-06-24 10:20:30.123+03
(4 rows)

```

Reply:
Thanks for your suggestions. I will add more test cases, and also validate the implementation using the whole set of possible datetime/timestamp column types.

Reply:
> We need to use `timestamp WITH time zone` as the column data type to make the conversion.

I see. That will probably need implementing a dedicated type, because `sa.DateTime` will just render a `TIMESTAMP` column, which would effectively be **WITHOUT time zone**?

Would that standard mapping, `sa.DateTime` -> `TIMESTAMP`, actually needed to be adjusted to render as `timestamp WITH time zone`?

The documentation of SQLAlchemy's `DateTime` type [says](https://github.com/sqlalchemy/sqlalchemy/blob/rel_2_0_31/lib/sqlalchemy/sql/sqltypes.py#L721-L739): 

> For the time representation within the datetime type, some backends include additional options, such as timezone support and fractional seconds support.  For fractional seconds, use the dialect-specific datatype, such as :class:`.mysql.TIME`.  For timezone support, use at least the :class:`_types.TIMESTAMP` datatype, if not the dialect-specific datatype object.

Reply:
I've just added 4482076ee51, aiming to improve the situation. However, test outcomes did not change in any way that would be relevant.

Reply:
619c991da2 starts using time zone `Europe/Kyiv` for testing, but it does not change anything on the test outcome.

Reply:
I've inspected the wire using `sudo ngrep -d lo0 -Wbyline port 4200`, and found that the JSON serialization does not take the time zone into account?
```json
{"stmt": "INSERT INTO foobar (name, date, datetime_notz, datetime_tz) VALUES (?, ?, ?, ?)", "args": ["foo", "2009-05-13", "2009-05-13T19:19:30.123456Z", "2009-05-13T19:19:30.123456Z"]}
```

Reply:
Contrary to my last statement, it looks like it works well when using direct SQL for querying.

### DDL
```sql
SHOW CREATE TABLE doc.foobar;
```
```sql
CREATE TABLE IF NOT EXISTS "doc"."foobar" (    
   "name" TEXT NOT NULL,                       
   "date" TIMESTAMP WITHOUT TIME ZONE,         
   "datetime_notz" TIMESTAMP WITHOUT TIME ZONE,
   "datetime_tz" TIMESTAMP WITH TIME ZONE,     
   PRIMARY KEY ("name")                        
)                                              
```

### DML
```sql
INSERT INTO foobar (name, date, datetime_notz, datetime_tz) VALUES ('foo', '2009-05-13', '2009-05-13T19:19:30.123456Z', '2009-05-13T19:19:30.123456Z');
INSERT INTO foobar (name, date, datetime_notz, datetime_tz) VALUES ('bar', '2009-05-13', '2009-05-13T19:19:30.123456+03:00', '2009-05-13T19:19:30.123456+03:00');
```

### DQL
```sql
SELECT * FROM foobar;
```
```
+------+---------------+---------------+---------------+
| name |          date | datetime_notz |   datetime_tz |
+------+---------------+---------------+---------------+
| foo  | 1242172800000 | 1242242370123 | 1242242370123 |
| bar  | 1242172800000 | 1242242370123 | 1242231570123 |
+------+---------------+---------------+---------------+
```

Reply:
In this spirit, it is clear that there is a flaw, and the relevant spot can be discovered easily. 💥

https://github.com/crate/sqlalchemy-cratedb/blob/df7fab91dfb947cafbc0ffa7e4c22c91317bacc1/src/sqlalchemy_cratedb/dialect.py#L125

Reply:
145f70154 aims to get the improvement straight, test outcomes now reflect the relevant conversion implicitly performed by CrateDB.